### DISCUSSION TITLE
How do I create additional annotations for a module that already has its own pyi file? (Stub resolution)

### DISCUSSION THREAD

**Discussion by argilya** [microsoft/pyright#3455] [Q&A] [answered]
Hello,

I would like to provide my own annotations for some of the functions in the `numpy` package, such as `linspace` and `arange`. Numpy has annotations for these functions, but they do not pass muster on `strict` mode. Pyright flags them as partially `Unknown`.

I understand that to create my own annotations, I need to create a `pyi` file at `typings/<package_name>/...`. For example, for `np.linspace` the file would be `typings/numpy/core/function_base.pyi`, and then the linspace annotation would simply be something like `def linspace(start: int, stop: int, step: int) -> List[int] or NDRange or something along those lines: ...`. I got this part to work.

That being said, when I create my own stubs, what happen's to numpy's existing stubs?

I notice that Numpy, in its source code, also has a file called `function_base.pyi`, where they have their own annotations of `linspace`, and a number of other functions like `logspace`, etc. Does my stub completely overwrite Numpy's stub? I think this is the case. Once I create my custom stub for `function_base.pyi` with just the `linspace` in it, Pyright sees my annotation for `linspace`, but loses annotations for all other functions in that module. So, for example, I lose annotations for `logspace` ,`geomspace`, etc.

Is there some obvious way to work around this? I want to create my own annotation for `linspace`, and only for `linspace`. I still want to have access to Numpy's (however flawed) annotations for the other functions in `function_base.pyi`. Is that possible to do?

PS

Ultimately, my question is about annotation resolution. How does Pyright decide which annotations it uses? I think it goes custom user-defined stubs, then in-line annotations in the package source, then `.pyi` files in the package, then external `<package>-stub` packages, then auto-generated annotations? Is that correct?

Thank you!

**Comment by erictraut** [comment]
Please refer to [this documentation](https://github.com/microsoft/pyright/blob/main/docs/import-resolution.md). Import resolution ordering for stubs is dictated by PEP 561, so you may find it a good resource as well.

**Comment by argilya** [comment, marked as answer]
Thank you, that was very helpful. To follow up on what I learned:

Is there a way to "merge" a custom stub file (in stubPath) with an in-line stub? Basically, I want to overwrite just a couple of function signatures that exist in the package's in-line stub file, without touching the rest. My solution so far has been to copy the _entire_ in-line stub file into my own stubPath, and then editing/adding things to it. Is there a more graceful way of doing it?

Thank you!

**Reply by erictraut** [reply]
Copying the entire stub is the right approach. Stub files are not designed to be "mergable".

**Reply by argilya** [reply]
Thank you!
